nshlib/nsh_usbconsole.c

```diff
@@ -88,15 +88,16 @@ static void nsh_configstdio(int fd, FAR struct console_stdio_s *pstate)
   dup2(fd, 1);
   dup2(fd, 2);
 
-  /* Setup the stdout */
-
-  pstate->cn_outfd     = 1;
-  pstate->cn_outstream = fdopen(1, "a");
-
-  /* Setup the stderr */
+  /* fdopen to get the stdin, stdout and stderr streams.
+   *
+   * fd = 0 is stdin  (read-only)
+   * fd = 1 is stdout (write-only, append)
+   * fd = 2 is stderr (write-only, append)
+   */
 
-  pstate->cn_errfd     = 2;
-  pstate->cn_errstream = fdopen(2, "a");
+  fdopen(0, "r");
+  fdopen(1, "a");
+  fdopen(2, "a");
```

Comment:
This is generally not needed.  The fact that fdopen() corrects the problem is an artifact of the fact that CONFIG_DEV_CONSOLE is not set in the configuration.

See sched/group/goupr_setupidlefiles.c:

     98 #ifdef CONFIG_DEV_CONSOLE
     99   fd = nx_open("/dev/console", O_RDWR);
    100   if (fd == 0)
    101     {
    102       /* Successfully opened /dev/console as stdin (fd == 0) */
    103
    104       fs_dupfd2(0, 1);
    105       fs_dupfd2(0, 2);
    106     }
    125 #endif

I believe that it is the logic in sched/group/goupr_setupidlefiles.c.  Since stdin, stdout, and stderr must always be available, it is wrong to permit CONFIG_DEV_CONSOLE to not be set.

A better solution would be to use /dev/null for tstdin, stdout, and stderr if CONFIG_DEV_CONSOLE  is not set.

Reply:
but dev/null or dev/zero may disable too.

Reply:
> but dev/null or dev/zero may disable too.

Yes, it would have to e changed so that it cannot be disabled.  Not a big deal.

Reply:
But, I am curious that line 87-89 already duplicate fd to 0-2, why we still need call fdopen at line 98-100 again?

Reply:
> But, I am curious that line 87-89 already duplicate fd to 0-2, why we still need call fdopen at line 98-100 again?

The file descriptors are duplicated, but the the streams are still not opened due to the failure of fs_fdopen() when the IDLE thread was created.  Duplicating the file descriptor does not create the stream.

Reply:
> > But, I am curious that line 87-89 already duplicate fd to 0-2, why we still need call fdopen at line 98-100 again?
> 
> The file descriptors are duplicated, but the the streams are still not opened due to the failure of fs_fdopen() when the IDLE thread was created. Duplicating the file descriptor does not create the stream.
> 

Ok, I understand the situation now.

> This should not be done in the the OS (not the application) OR it could be done in _all_ applications and the non-standard requirement should be well documented.

Maybe the right thing we should do:

1. Always enable DEV_ZERO or DEV_NULL
2. Or skip check 0-2 valid in fs_fdopen

> The latter is basically the way things worked before your change that broke this.

But with the patch: https://github.com/apache/incubator-nuttx/pull/2263
I think the new behaviour is same as before now. If CONFIG_DEV_CONSOLE isn't enabled, fdopen is always required regardless whether my patch apply.